File: java/programmers/1단계/src/약수의개수와덧셈.java
import javax.imageio.ImageWriteParam;

public class 약수의개수와덧셈 {
    public int solution(int left, int right) {
        int answer = 0;
        for (int i = left; i <= right; i++) {
            if(divisor_even(i)){
                answer += i;
            } else{
                answer -= i;
            }
        }
        return answer;
    }
    private boolean divisor_even(int num) {
        int cnt = 0;
        for (int i = 1; i <= num; i++) {
            if (num % i == 0) {
                cnt++;
            }
        }
        return cnt % 2 == 0;
    }

    /**
     * 제곱수의 약수의 개수는 홀수이다.
     */
    public int solution2(int left, int right) {
        int answer = 0;
        for (int i = left; i <= right; i++) {
            // 제곱수인 경우 약수의 개수는 홀수
            if (i % Math.sqrt(i) == 0) {
                answer -= i;
            }
            // 제곱수가 아닌 경우 약수의 개수는 짝수
            else answer += i;
        }
        return answer;
    }

    public static void main(String[] args) {
        약수의개수와덧셈 sol = new 약수의개수와덧셈();
        int left = 24;
        int right = 27;
        System.out.println(sol.solution(left, right));
        System.out.println(sol.solution2(left, right));
    }
}

File: python/python_Programmers/1단계/약수의개수와덧셈.py
# 내풀이
def solution(left, right):
    arr = []
    answer = 0
    for num in range(left, right+1):

        for i in range(1, num+1):
            if num % i == 0:
                arr.append(i)
        if len(arr) % 2 == 0:
            answer += num
        else:
            answer -= num
        arr = []
    return answer
print(solution(24, 27))

# 모범 풀이 ( 제곱수의 약수의 갯수는 항상 홀수이다.)
def solution2(left, right):
    answer = 0
    for i in range(left,right+1):
        if int(i**0.5)==i**0.5:
            answer -= i
        else:
            answer += i
    return answer

import math

def solution3(left, right):
    answer = 0
    for i in range(left, right + 1, 1):
        sqrt = math.sqrt(i)
        if int(sqrt) == sqrt:
            answer -= i
        else:
            answer += i

    return answer